code/game/turfs/unsimulated/snow.dm

```diff
@@ -108,119 +170,175 @@
+				A.relativewall()
+
+/obj/dirtpath/relativewall()
+	overlays.Cut()
+	if(!diags.len)
+		for(var/diagdir in diagonal)
+			diags["diag[diagdir]"] = image('icons/turf/new_snow.dmi', "permafrost_corner", dir = diagdir)
 	var/junction = findSmoothingNeighbors()
 	var/dircount = 0
 	for(var/direction in diagonal)
-		if(istype(get_step(src, direction), /turf/unsimulated/floor/snow/permafrost))
+		var/turf/adj_tile = get_step(src, direction)
+		if(isSmoothableNeighbor(adj_tile) || locate(/obj/dirtpath) in adj_tile)
 			if((direction & junction) == direction)
-				overlays += dirt_layers["diag[direction]"]
+				overlays += diags["diag[direction]"]
 				dircount++
-	if(dircount == 4)
-		overlays.Cut()
-		icon_state = "permafrost_full"
-		overlays += snow_layers["1"]
-		overlays += snow_layers["2"]
-	else if(junction)
-		overlays += dirt_layers["snow[junction]"]
+
+	switch(dircount)
+		if(4)
+			overlays.Cut()
+			junction = "_full"
+		if(0)
+			if(!junction)
+				junction = "_circle"
+	icon_state = "permafrost[junction]"
+
+/turf/snow/ice
+	icon_state = "ice"
+
+/turf/snow/ice/New()
+	..()
+	snowballs = SNOWBALLS_ICE
+	new /obj/glacier(src)
+
+/obj/glacier
+	name = "glacier"
+	desc = "A frozen lake, kept solid by temperatures way below freezing."
+	icon = 'icons/turf/ice.dmi'
+	icon_state = "ice1"
+	canSmoothWith = "/obj/glacier"
+	anchored = 1
+	density = 0
+	plane = PLATING_PLANE
+	var/isedge
+	var/hole = 0
+
+/obj/glacier/New(var/icon_update_later = 0)
+	var/turf/snow/T = loc
+	if(!istype(T))
+		qdel(src)
+		return
+	..()
+	T.snowballs = SNOWBALLS_ICE
+	if(icon_update_later)
+		relativewall()
+		relativewall_neighbours()
+
+/obj/glacier/relativewall()
+	overlays.Cut()
+	var/junction = 0
+	isedge = 0
+	var/edgenum = 0
+	var/edgesnum = 0
+	for(var/direction in alldirs)
+		var/turf/adj_tile = get_step(src, direction)
+		var/obj/glacier/adj_glacier = locate(/obj/glacier) in adj_tile
+		if(adj_glacier)
+			junction |= dir_to_smoothingdir(direction)
+			if(adj_glacier.isedge && direction in cardinal)
+				edgenum |= direction
+				edgesnum = adj_glacier.isedge
+	if(junction == SMOOTHING_ALLDIRS) // you win the not-having-to-smooth-lotterys
+		icon_state = "ice[rand(1,6)]"
 	else
-		overlays += dirt_layers["snow0"]
+		switch(junction)
+			if(SMOOTHING_L_CURVES)
+				isedge = junction
+				relativewall_neighbours()
+		icon_state = "junction[junction]"
+	if(edgenum && !isedge)
+		icon_state = "edge[edgenum]-[edgesnum]"
+
+	if(hole)
+		overlays += image(icon,"hole_overlay")
+
+/obj/glacier/relativewall_neighbours()
+	..()
+	for(var/direction in diagonal)
+		var/turf/adj_tile = get_step(src, direction)
+		if(isSmoothableNeighbor(adj_tile))
+			adj_tile.relativewall()
+		for(var/atom/A in adj_tile)
+			if(isSmoothableNeighbor(A))
+				A.relativewall()
+
+/obj/glacier/can_fish()
+	return hole
+
+/obj/glacier/attackby(var/obj/item/W, mob/user)
+	to_chat(user,"<span class='notice'>you smash a hole in the ice with \the [W]</span>") // todo: better
```

Comment:
I did warn Sood that there were aspects that needed a little more work on, this is one of then